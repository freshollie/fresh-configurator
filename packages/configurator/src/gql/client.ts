import { InMemoryCache, ApolloClient, gql } from "@apollo/client";
import WebSocketLink from "./WebSocketLink";
import { Resolvers, Configurator } from "./__generated__";
import introspection from "./__generated__/introspection.json";
import { versionInfo } from "../util";
import {
  LogsDocument,
  SelectedTabDocument,
  ConnectionSettingsDocument,
} from "./queries/Configurator.graphql";

const typeDefs = gql`
  type Query {
    configurator: Configurator!
  }

  type Mutation {
    setConnectionSettings(port: String!, baudRate: Int): Boolean
    setConnection(connection: ID): Boolean
    setConnecting(value: Boolean!): Boolean
    setTab(tabId: String!): Boolean
    setExpertMode(enabled: Boolean!): Boolean
    log(message: String!): Boolean
  }

  type Configurator {
    port: String
    connection: ID
    connecting: Boolean!
    baudRate: Int!
    tab: String
    expertMode: Boolean!
    logs: [Log!]!
  }

  type Log {
    time: String!
    message: String!
  }
`;

export const cache = (): InMemoryCache =>
  new InMemoryCache({
    typePolicies: Object.fromEntries(
      // eslint-disable-next-line no-underscore-dangle
      introspection.__schema.types
        .filter(({ kind }) => kind === "OBJECT")
        .map(({ name }) => [name, { merge: true }])
    ),
  });

export const resolvers = (initialState?: {
  connecting?: boolean;
  connection?: string | null;
  port?: string;
  baudRate?: number;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
}): any => {
  const config: Resolvers = {
    Query: {
      configurator: () => ({
        __typename: "Configurator",

        port: "",
        baudRate: 115200,
        connecting: false,
        connection: null,
        expertMode: false,
        logs: [],
        ...initialState,
      }),
    },
    Mutation: {
      setTab: (_, { tabId }, { client }) => {
        client.writeQuery({
          query: SelectedTabDocument,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              tab: tabId,
            },
          },
        });
        return null;
      },
      setConnectionSettings: (_, { port, baudRate }, { client }) => {
        const existingData = {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          ...client.readQuery({
            query: ConnectionSettingsDocument,
          })!.configurator,
        };

        if (typeof baudRate === "number") {
          existingData.baudRate = baudRate;
        }
        existingData.port = port;

        client.writeQuery({
          query: ConnectionSettingsDocument,
          data: {
            __typename: "Query",
            configurator: {
              ...existingData,
            },
          },
        });

        return null;
      },
      setConnecting: (_, { value }, { client }) => {
        client.writeQuery({
          query: gql`
            query {
              configurator {
                connecting
              }
            }
          `,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              connecting: value,
            },
          },
        });

        return null;
      },
      setConnection: (_, { connection }, { client }) => {
        client.writeQuery({
          query: gql`
            query {
              configurator {
                connection
              }
            }
          `,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              connection,
            },
          },
        });

        return null;
      },
      // setExpertMode: (_, { enabled }, { client }) => {
      //   client.writeQuery({
      //     document: Ex,
      //     data: {
      //       __typename: "Query",
      //       configurator: {
      //         __typename: "Configurator",
      //         expertMode: enabled,
      //       } as Configurator,
      //     },
      //   });

      //   return null;
      // },
      log: (_, { message }, { client }) => {
        const logs =
          client.readQuery({
            query: LogsDocument,
          })?.configurator.logs ?? [];

        client.writeQuery({
          query: LogsDocument,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              logs: logs.concat([
                {
                  time: new Date().toISOString(),
                  message,
                  __typename: "Log" as const,
                },
              ]),
            },
          },
        });

        return null;
      },
    },
  };
  return config;
};

// extract the backend address from the URL search query, as this can
// be dynamically passed to us by electron
const searchParams = new URLSearchParams(window.location.search.slice(1));
const BACKEND = searchParams.get("backend") ?? "ws://localhost:9000";

const client = new ApolloClient({
  cache: cache(),
  typeDefs,
  resolvers: resolvers(),
  link: new WebSocketLink({
    url: `${BACKEND}/graphql`,
    keepAlive: 99999999999,
  }),
});

const writeInitial = (): void => {
  const { os, version, chromeVersion } = versionInfo();
  client.writeQuery({
    query: gql`
      query {
        configurator {
          port
          baudRate
          tab
          expertMode
          connecting
          connection
          logs {
            time
            message
          }
        }
      }
    `,
    data: {
      __typename: "Query",
      configurator: {
        port: null,
        baudRate: 115200,
        tab: null,
        expertMode: false,
        connecting: false,
        connection: null,
        logs: [
          {
            time: new Date().toISOString(),
            message: `Running - OS: <strong>${os}</strong>, Chrome: <strong>${chromeVersion}</strong>, Configurator: <strong>${version}</strong>`,
            __typename: "Log",
          },
        ],
        __typename: "Configurator",
      } as Configurator,
    },
  });
};

writeInitial();
client.onResetStore(() => Promise.resolve(writeInitial()));

export default client;
