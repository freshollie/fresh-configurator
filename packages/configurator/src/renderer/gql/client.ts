import {
  InMemoryCache,
  ApolloClient,
  ApolloLink,
  gql as schemaTypes,
  NormalizedCacheObject,
} from "@apollo/client";
import { Resolvers } from "./__generated__/schema";
import introspection from "./__generated__/introspection.json";
import { versionInfo } from "../util";
import IpcLink from "./links/IpcLink";

import { gql } from "./apollo";
import config from "../config";
import WebWorkerLink from "./links/WebWorkerLink";

const typeDefs = schemaTypes`
  type Query {
    configurator: Configurator!
  }

  type Mutation {
    log(message: String!): Boolean
  }

  type Configurator {
    logs: [Log!]!
  }

  type Log {
    time: String!
    message: String!
  }
`;

const Logs = gql(/* GraphQL */ `
  query Logs {
    configurator @client {
      logs {
        time
        message
      }
    }
  }
`);

// extract the backend address from the URL search query, as this can
// be dynamically passed to us by electron

export const artifactsAddress =
  config.artifactsFolder ??
  `${
    // Something about esbuild makes optional chaining not an option here :/
    (config.wsBackend ? config.wsBackend.replace("ws", "http") : undefined) ??
    "http://localhost:9000"
  }/job-artifacts`;

const createRequiredLink = async (): Promise<ApolloLink> => {
  if (window.ipcRenderer) {
    return new IpcLink({ ipc: window.ipcRenderer });
  }

  if (config.wsBackend) {
    const { default: WebSocketLink } = await import("./links/WebSocketLink");
    return new WebSocketLink({
      url: `${config.wsBackend}/graphql`,
      keepAlive: 99999999999,
    });
  }

  // Hack to get around ts-jest trying to compile `import.meta.url`
  const { default: schemaExecutor } = await import(
    "../../worker/SchemaExecutor.bootstrap"
  );
  const mocked = config.isMocked;

  const worker = await schemaExecutor.initialise(mocked);
  return new WebWorkerLink(worker);
};

export const createClient = async (): Promise<
  ApolloClient<NormalizedCacheObject>
> => {
  const cache = (): InMemoryCache =>
    new InMemoryCache({
      typePolicies: Object.fromEntries(
        // eslint-disable-next-line no-underscore-dangle
        introspection.__schema.types
          .filter(({ kind }) => kind === "OBJECT")
          .map(({ name }) => [name, { merge: true }])
      ),
    });

  const resolvers = (): Resolvers => ({
    Query: {
      configurator: () => ({
        __typename: "Configurator",
        logs: [],
      }),
    },
    Mutation: {
      log: (_, { message }, { client }) => {
        const logs =
          client.readQuery({
            query: Logs,
          })?.configurator.logs ?? [];

        client.writeQuery({
          query: Logs,
          data: {
            __typename: "Query",
            configurator: {
              __typename: "Configurator",
              logs: logs.concat([
                {
                  time: new Date().toISOString(),
                  message,
                  __typename: "Log" as const,
                },
              ]),
            },
          },
        });

        return null;
      },
    },
  });

  const client = new ApolloClient({
    cache: cache(),
    typeDefs,
    resolvers: resolvers() as never,
    link: ApolloLink.from([
      // new RetryLink({
      //   // Retry all operations unless retry is explicitly disabled
      //   attempts: (_, operation) => operation.getContext().retry !== false,
      //   delay: {
      //     initial: 10,
      //     max: 5,
      //     jitter: true,
      //   },
      // }),
      // new ApolloLink((operation, forward) =>
      //   forward(operation).map((data) => {
      //     if (data.errors?.[0]) {
      //       console.log("test");
      //       throw new Error(`${data.errors[0].message}`);
      //     }
      //     return data;
      //   })
      // ),
      await createRequiredLink(),
    ]),
  });

  const writeInitial = (): void => {
    const { os, version, chromeVersion } = versionInfo();
    client.writeQuery({
      query: gql(/* GraphQL */ `
        query InitWrite {
          configurator {
            logs {
              time
              message
            }
          }
        }
      `),
      data: {
        __typename: "Query",
        configurator: {
          logs: [
            {
              time: new Date().toISOString(),
              message: `Running - OS: <strong>${os}</strong>, Chrome: <strong>${chromeVersion}</strong>, Configurator: <strong>${version}</strong>`,
              __typename: "Log",
            },
          ],
          __typename: "Configurator",
        },
      },
    });
  };

  writeInitial();
  client.onResetStore(() => Promise.resolve(writeInitial()));

  return client;
};
