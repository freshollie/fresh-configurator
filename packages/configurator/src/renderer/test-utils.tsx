import React from "react";
import { render as renderFunction, RenderResult } from "@testing-library/react";
import { Provider } from "bumbag";
import theme from "./theme";

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: (query: string) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(), // Deprecated
    removeListener: jest.fn(), // Deprecated
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  }),
});

const AllTheProviders: React.FC = ({ children }) => (
  <Provider theme={theme}>{children}</Provider>
);

const customRender = (
  ui: Parameters<typeof renderFunction>[0],
  options?: Parameters<typeof renderFunction>[1]
): RenderResult => renderFunction(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from "@testing-library/react";

// override render method
export { customRender as render };
