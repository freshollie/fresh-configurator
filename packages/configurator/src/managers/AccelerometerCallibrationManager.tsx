import React from "react";
import { CallibrateAccelerometerDocument } from "../gql/mutations/Device.graphql";
import Button from "../components/Button";
import useConnectionState from "../hooks/useConnectionState";
import useLogger from "../hooks/useLogger";
import { useMutation } from "../gql/apollo";

const AccelerometerCallibrationManager: React.FC = () => {
  const { connection } = useConnectionState();
  const log = useLogger();
  const [calibrate, { loading }] = useMutation(
    CallibrateAccelerometerDocument,
    {
      variables: {
        connection: connection ?? "",
      },
      onCompleted: () => {
        log(
          'Accelerometer calibration <span class="message-positive">finished</span>'
        );
      },
      onError: (e) => {
        log(`Error callibrating accelerometer: ${e.message}`);
      },
    }
  );

  return (
    <Button
      data-testid="calibrate-acc-button"
      disabled={loading || !connection}
      onClick={() => {
        calibrate();
        log("Accelerometer calibration started");
      }}
    >
      {loading ? "Calibrating..." : "Calibrate Accelerometer"}
    </Button>
  );
};

export default AccelerometerCallibrationManager;
