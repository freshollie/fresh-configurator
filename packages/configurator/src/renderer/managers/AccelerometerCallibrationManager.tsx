import React from "react";
import { Button } from "bumbag";
import useConnection from "../hooks/useConnection";
import useLogger from "../hooks/useLogger";
import { gql, useMutation } from "../gql/apollo";

const AccelerometerCallibrationManager: React.FC = () => {
  const connection = useConnection();
  const log = useLogger();
  const [calibrate, { loading }] = useMutation(
    gql(/* GraphQL */ `
      mutation CallibrateAccelerometer($connection: ID!) {
        deviceCallibrateAccelerometer(connectionId: $connection)
      }
    `),
    {
      variables: {
        connection,
      },
      onCompleted: () => {
        log(
          'Accelerometer calibration <span class="message-positive">finished</span>'
        );
      },
      onError: (e) => {
        log(`Error callibrating accelerometer: ${e.message}`);
      },
    }
  );

  return (
    <Button
      data-testid="calibrate-acc-button"
      disabled={loading || !connection}
      onClick={() => {
        calibrate();
        log("Accelerometer calibration started");
      }}
    >
      {loading ? "Calibrating..." : "Calibrate Accelerometer"}
    </Button>
  );
};

export default AccelerometerCallibrationManager;
