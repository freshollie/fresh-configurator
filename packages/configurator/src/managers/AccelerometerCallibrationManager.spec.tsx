import React from "react";
import { MockedProvider, MockedResponse } from "@apollo/client/testing";

import { FetchResult } from "@apollo/client";
import { render, fireEvent, waitFor } from "../test-utils";
import AccelerometerCallibrationManager from "./AccelerometerCallibrationManager";
import {
  CallibrateAccelerometerDocument,
  CallibrateAccelerometerMutation,
} from "../gql/mutations/Device.graphql";

const mockConnectionId = "someconnectionid";

const log = jest.fn();

const clientResolvers = {
  Query: {
    configurator: () => ({
      connection: mockConnectionId,
      connecting: false,
      __typename: "Configurator",
    }),
  },
  Mutation: {
    log,
  },
};

const mockCallibrateMutation = (
  connection: string,
  callback: () => void
): MockedResponse => ({
  request: {
    query: CallibrateAccelerometerDocument,
    variables: {
      connection,
    },
  },
  result: (): FetchResult<CallibrateAccelerometerMutation> => {
    callback();
    return {
      data: {
        __typename: "Mutation",
        deviceCallibrateAccelerometer: null,
      },
    };
  },
  delay: 100,
});

describe("AccelerometerCallibrationManager", () => {
  it("should callibrate the device accelerometer when button is clicked", async () => {
    const callibrate = jest.fn();
    const { getByTestId } = render(
      <MockedProvider
        resolvers={clientResolvers}
        mocks={[mockCallibrateMutation(mockConnectionId, callibrate)]}
      >
        <AccelerometerCallibrationManager />
      </MockedProvider>
    );

    await waitFor(() =>
      expect(getByTestId("calibrate-acc-button")).toBeEnabled()
    );

    expect(getByTestId("calibrate-acc-button")).toMatchSnapshot();
    fireEvent.click(getByTestId("calibrate-acc-button"));

    await waitFor(() =>
      expect(getByTestId("calibrate-acc-button")).toBeDisabled()
    );
    expect(getByTestId("calibrate-acc-button")).toMatchSnapshot();

    await waitFor(() =>
      expect(getByTestId("calibrate-acc-button")).toBeEnabled()
    );
    expect(getByTestId("calibrate-acc-button")).toMatchSnapshot();
    expect(callibrate).toHaveBeenCalled();
  });
});
