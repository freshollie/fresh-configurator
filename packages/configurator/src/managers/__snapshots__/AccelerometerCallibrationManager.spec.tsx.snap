// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccelerometerCallibrationManager should callibrate the device accelerometer when button is clicked 1`] = `
<button
  class="css-1k04dns css-63it0p css-jwhr8m bb-Box bb-Button"
  data-testid="calibrate-acc-button"
  style=""
  type="button"
>
  Calibrate Accelerometer
</button>
`;

exports[`AccelerometerCallibrationManager should callibrate the device accelerometer when button is clicked 2`] = `
<button
  aria-disabled="true"
  class="css-kkyqvo css-63it0p css-jwhr8m bb-Box bb-Button"
  data-testid="calibrate-acc-button"
  disabled=""
  style="pointer-events: none;"
  type="button"
>
  Calibrating...
</button>
`;

exports[`AccelerometerCallibrationManager should callibrate the device accelerometer when button is clicked 3`] = `
<button
  class="css-1k04dns css-63it0p css-jwhr8m bb-Box bb-Button"
  data-testid="calibrate-acc-button"
  style=""
  type="button"
>
  Calibrate Accelerometer
</button>
`;
